blockly-rtc/src/MarkerManager.js

```diff
@@ -0,0 +1,149 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Wrapper class for interacting with the Blockly MarkerManager.
+ * @author navil@google.com (Navil Perez)
+ */
+
+import * as Blockly from 'blockly/dist';
+import MarkerUpdate from './MarkerUpdate';
+
+export default class MarkerManager {
+  constructor(workspaceId, sendMarkerUpdate, getMarkerUpdates,
+      addClient, getBroadcastMarkerUpdates) {
+    this.workspaceId = workspaceId
+    this.colours =  [
+        '#fcba03', '#03fc20', '#03f0fc', '#035efc', '#5603fc', '#fc03d2'];
+    this.sendMarkerUpdate = sendMarkerUpdate;
+    this.getMarkerUpdates = getMarkerUpdates;
+    this.addClient = addClient;
+    this.broadcastMarkerHandler = getBroadcastMarkerUpdates;
+  };
+
+  /**
+   * Initialize the workspace by creating and registering markers for all active
+   * users and activating the handling of recieving MarkerUpdates from the server.
+   * @public
+   */
+  async initMarkers() {
+    await this.addClient(this.workspaceId);
+    const markerUpdates = await this.getMarkerUpdates();
+    markerUpdates.forEach((markerUpdate) => {
+      this.createMarker_(markerUpdate);
+      if (this.broadcastMarkerHandler) {
+        this.broadcastMarkerHandler(this.updateMarkerLocations_.bind(this));
+      } else {
+        this.pollServer_();
+      };
+    });
+  };
+
+  /**
+   * Create a MarkerUpdate from a Blockly event and send it to the server.
+   * @param {!Blockly.Event} The event for which to create a MarkerUpdate.
+   * @public
+   */
+  async handleEvent(event) {
+    const markerUpdate = MarkerUpdate.fromEvent(event);
+    await this.sendMarkerUpdate(markerUpdate);
+  };
+
+  /**
+   * Periodically query the database for MarkerUpdates.
+   * @private
+   */
+  async pollServer_() {
+    const markerUpdates = await this.getMarkerUpdates();
+    await this.updateMarkerLocations_(markerUpdates);
+    setTimeout(() => {
+      this.pollServer_();
+    }, 5000)
+  };
+
+  /**
+   * Get the workspace with the id specified by the MarkerUpdate.
+   * @return {Blockly.Workspace} The sought after workspace or null if not found.
+   * @private
+   */  
+  getWorkspace_() {
+    return Blockly.Workspace.getById(this.workspaceId);
+  };
+
+  /**
+   * Get the MarkerManager for the workspace specified by MarkerUpdate.
+   * @return {Blockly.MarkerManager} The sought after MarkerManager or null if not
+   * found.
+   * @private
+   */  
+  getMarkerManager_() {
+    return this.getWorkspace_().getMarkerManager();
```

Comment:
This will break if `getWorkspace_()` returned `null`.

Reply:
Done.